bin/teamcity-e2e-slack-notify.mjs

```diff
@@ -0,0 +1,146 @@
+import fs from 'fs';
+import { exit } from 'process';
+import fetch from 'node-fetch';
+import yargs from 'yargs';
+
+/* CLI Arguments */
+
+const { argv } = yargs( process.argv ).options( {
+	file: {
+		type: 'string',
+		required: true,
+		describe: 'Path to a JSON file produced by Jest',
+	},
+} );
+
+const filePath = argv.file;
+
+/* Methods */
+
+/**
+ * Opens and parses the test result JSON produced by Jest.
+ * @returns {JSON} JSON parsed test suite result.
+ * @throws {Error} If file was not present.
+ */
+function openJestOutputJSON() {
+	let data;
+	try {
+		fs.accessSync( filePath );
+		data = fs.readFileSync( filePath, 'utf8' );
+	} catch ( error ) {
+		console.error( 'An error occurred while accessing the file:', error );
+		exit();
+	}
+	return JSON.parse( data );
+}
+
+/**
+ * Extracts all stack traces for failing test steps.
+ * @param {JSON} results Test suite results produced by Jest.
+ * @returns {{step: string, error: string}[]} Array of failure objects.
+ */
+function extractFailureMessages( results ) {
+	const failureMessages = [];
+
+	// Short circuit if nothing consistently failed.
+	// Consistently here means failed on both initial run and then the retry.
+	if ( ! results.numFailedTests ) {
+		return failureMessages;
+	}
+
+	for ( const result of results.testResults ) {
+		for ( const step of result.assertionResults ) {
+			if ( step.failureMessages.length !== 0 ) {
+				failureMessages.push( { step: step.ancestorTitles, error: step.failureMessages } );
+			}
+		}
+	}
```

Comment:
The general structure of the Jest-produced JSON is as follows:

<details>
```
{
    "numFailedTestSuites": 2,
    "numFailedTests": 2,
    "numPassedTestSuites": 0,
    "numPassedTests": 3,
    "numPendingTestSuites": 0,
    "numPendingTests": 7,
    "numRuntimeErrorTestSuites": 0,
    "numTodoTests": 0,
    "numTotalTestSuites": 2,
    "numTotalTests": 12,
    "openHandles": [],
    "snapshot": {
        "added": 0,
        "didUpdate": false,
        "failure": false,
        "filesAdded": 0,
        "filesRemoved": 0,
        "filesRemovedList": [],
        "filesUnmatched": 0,
        "filesUpdated": 0,
        "matched": 0,
        "total": 0,
        "unchecked": 0,
        "uncheckedKeysByFile": [],
        "unmatched": 0,
        "updated": 0
    },
    "startTime": 1698224361446,
    "success": false,
    "testResults": [
        {
            "assertionResults": [
                {
                    "ancestorTitles": [
                        "Help Center: Interact with Results",
                        "Search for Help article"
                    ],
                    "duration": 68,
                    "failureDetails": [],
                    "failureMessages": [],
                    "fullName": "Help Center: Interact with Results Search for Help article Open Help Center",
                    "invocations": 1,
                    "location": null,
                    "numPassingAsserts": 0,
                    "retryReasons": [],
                    "status": "passed",
                    "title": "Open Help Center"
                },
                {
                    "ancestorTitles": [
                        "Help Center: Interact with Results",
                        "Search for Help article"
                    ],
                    "duration": 11913,
                    "failureDetails": [
                        {
                            "name": "TimeoutError"
                        }
                    ],
                    "failureMessages": [
                        "locator.waitFor: Timeout 10000ms exceeded.\n=========================== logs ===========================\nwaiting for locator('lkwelrer') to be visible\n============================================================\n    at Object.waitFor (/Users/edwintakahashi/workspace/wp-calypso/test/e2e/specs/help-center/help-center__general-interaction.ts:38:37)"
                    ],
                    "fullName": "Help Center: Interact with Results Search for Help article Search for posts-related help article",
                    "invocations": 1,
                    "location": null,
                    "numPassingAsserts": 0,
                    "retryReasons": [],
                    "status": "failed",
                    "title": "Search for posts-related help article"
                },
                {
                    "ancestorTitles": [
                        "Help Center: Interact with Results",
                        "Search for Help article"
                    ],
                    "duration": null,
                    "failureDetails": [],
                    "failureMessages": [],
                    "fullName": "Help Center: Interact with Results Search for Help article Click on the second Help Docs result",
                    "invocations": 1,
                    "location": null,
                    "numPassingAsserts": 0,
                    "retryReasons": [],
                    "status": "pending",
                    "title": "Click on the second Help Docs result"
                },
                {
                    "ancestorTitles": [
                        "Help Center: Interact with Results",
                        "Search for Help article"
                    ],
                    "duration": null,
                    "failureDetails": [],
                    "failureMessages": [],
                    "fullName": "Help Center: Interact with Results Search for Help article Help Doc article is shown",
                    "invocations": 1,
                    "location": null,
                    "numPassingAsserts": 0,
                    "retryReasons": [],
                    "status": "pending",
                    "title": "Help Doc article is shown"
                },
                {
                    "ancestorTitles": [
                        "Help Center: Interact with Results",
                        "Navigate to Calypso Link"
                    ],
                    "duration": null,
                    "failureDetails": [],
                    "failureMessages": [],
                    "fullName": "Help Center: Interact with Results Navigate to Calypso Link Close article and return to search results",
                    "invocations": 1,
                    "location": null,
                    "numPassingAsserts": 0,
                    "retryReasons": [],
                    "status": "pending",
                    "title": "Close article and return to search results"
                },
                {
                    "ancestorTitles": [
                        "Help Center: Interact with Results",
                        "Navigate to Calypso Link"
                    ],
                    "duration": null,
                    "failureDetails": [],
                    "failureMessages": [],
                    "fullName": "Help Center: Interact with Results Navigate to Calypso Link Clear search results",
                    "invocations": 1,
                    "location": null,
                    "numPassingAsserts": 0,
                    "retryReasons": [],
                    "status": "pending",
                    "title": "Clear search results"
                },
                {
                    "ancestorTitles": [
                        "Help Center: Interact with Results",
                        "Navigate to Calypso Link"
                    ],
                    "duration": null,
                    "failureDetails": [],
                    "failureMessages": [],
                    "fullName": "Help Center: Interact with Results Navigate to Calypso Link Search for \"domain\"",
                    "invocations": 1,
                    "location": null,
                    "numPassingAsserts": 0,
                    "retryReasons": [],
                    "status": "pending",
                    "title": "Search for \"domain\""
                },
                {
                    "ancestorTitles": [
                        "Help Center: Interact with Results",
                        "Navigate to Calypso Link"
                    ],
                    "duration": null,
                    "failureDetails": [],
                    "failureMessages": [],
                    "fullName": "Help Center: Interact with Results Navigate to Calypso Link Click on the first Calypso Link result",
                    "invocations": 1,
                    "location": null,
                    "numPassingAsserts": 0,
                    "retryReasons": [],
                    "status": "pending",
                    "title": "Click on the first Calypso Link result"
                },
                {
                    "ancestorTitles": [
                        "Help Center: Interact with Results",
                        "Navigate to Calypso Link"
                    ],
                    "duration": null,
                    "failureDetails": [],
                    "failureMessages": [],
                    "fullName": "Help Center: Interact with Results Navigate to Calypso Link Calypso Link opens in a new page",
                    "invocations": 1,
                    "location": null,
                    "numPassingAsserts": 0,
                    "retryReasons": [],
                    "status": "pending",
                    "title": "Calypso Link opens in a new page"
                }
            ],
            "endTime": 1698224382611,
            "message": "\\u001b[1m\\u001b[31m  \\u001b[1m● \\u001b[22m\\u001b[1mHelp Center: Interact with Results › Search for Help article › Search for posts-related help article\\u001b[39m\\u001b[22m\n\n    locator.waitFor: Timeout 10000ms exceeded.\n    =========================== logs ===========================\n    waiting for locator('lkwelrer') to be visible\n    ============================================================\n\\u001b[2m\\u001b[22m\n\\u001b[2m    \\u001b[0m \\u001b[90m 36 |\\u001b[39m    \\u001b[90m// Until that's fixed, let's steer clear and search a different topic.\\u001b[39m\\u001b[0m\\u001b[22m\n\\u001b[2m    \\u001b[0m \\u001b[90m 37 |\\u001b[39m    \\u001b[36mawait\\u001b[39m supportComponent\\u001b[33m.\\u001b[39msearch( \\u001b[32m'posts'\\u001b[39m )\\u001b[33m;\\u001b[39m\\u001b[0m\\u001b[22m\n\\u001b[2m    \\u001b[0m\\u001b[31m\\u001b[1m>\\u001b[22m\\u001b[2m\\u001b[39m\\u001b[90m 38 |\\u001b[39m    \\u001b[36mawait\\u001b[39m page\\u001b[33m.\\u001b[39mlocator( \\u001b[32m'lkwelrer'\\u001b[39m )\\u001b[33m.\\u001b[39mwaitFor()\\u001b[33m;\\u001b[39m\\u001b[0m\\u001b[22m\n\\u001b[2m    \\u001b[0m \\u001b[90m    |\\u001b[39m                                     \\u001b[31m\\u001b[1m^\\u001b[22m\\u001b[2m\\u001b[39m\\u001b[0m\\u001b[22m\n\\u001b[2m    \\u001b[0m \\u001b[90m 39 |\\u001b[39m   } )\\u001b[33m;\\u001b[39m\\u001b[0m\\u001b[22m\n\\u001b[2m    \\u001b[0m \\u001b[90m 40 |\\u001b[39m\\u001b[0m\\u001b[22m\n\\u001b[2m    \\u001b[0m \\u001b[90m 41 |\\u001b[39m   it( \\u001b[32m'Click on the second Help Docs result'\\u001b[39m\\u001b[33m,\\u001b[39m \\u001b[36masync\\u001b[39m \\u001b[36mfunction\\u001b[39m () {\\u001b[0m\\u001b[22m\n\\u001b[2m\\u001b[22m\n\\u001b[2m      \\u001b[2mat Object.waitFor (\\u001b[22m\\u001b[2m\\u001b[0m\\u001b[36mspecs/help-center/help-center__general-interaction.ts\\u001b[39m\\u001b[0m\\u001b[2m:38:37)\\u001b[22m\\u001b[2m\\u001b[22m\n",
            "name": "/Users/edwintakahashi/workspace/wp-calypso/test/e2e/specs/help-center/help-center__general-interaction.ts",
            "startTime": 1698224361875,
            "status": "failed",
            "summary": ""
        },
        {
            "assertionResults": [
                {
                    "ancestorTitles": [
                        "Server-side Rendering",
                        "Check SSR endoint: log-in"
                    ],
                    "duration": 1777,
                    "failureDetails": [],
                    "failureMessages": [],
                    "fullName": "Server-side Rendering Check SSR endoint: log-in Check SSR: log-in",
                    "invocations": 1,
                    "location": null,
                    "numPassingAsserts": 0,
                    "retryReasons": [],
                    "status": "passed",
                    "title": "Check SSR: log-in"
                },
                {
                    "ancestorTitles": [
                        "Server-side Rendering",
                        "Check SSR endoint: themes"
                    ],
                    "duration": 3121,
                    "failureDetails": [],
                    "failureMessages": [],
                    "fullName": "Server-side Rendering Check SSR endoint: themes Check SSR: themes",
                    "invocations": 1,
                    "location": null,
                    "numPassingAsserts": 0,
                    "retryReasons": [],
                    "status": "passed",
                    "title": "Check SSR: themes"
                },
                {
                    "ancestorTitles": [
                        "Server-side Rendering",
                        "Check SSR endoint: theme/twentytwentythree"
                    ],
                    "duration": 6341,
                    "failureDetails": [{}],
                    "failureMessages": [
                        "Error: \n    at Object.<anonymous> (/Users/edwintakahashi/workspace/wp-calypso/test/e2e/specs/infrastructure/infrastructure__ssr-enabled.ts:24:12)"
                    ],
                    "fullName": "Server-side Rendering Check SSR endoint: theme/twentytwentythree Check SSR: theme/twentytwentythree",
                    "invocations": 1,
                    "location": null,
                    "numPassingAsserts": 0,
                    "retryReasons": [],
                    "status": "failed",
                    "title": "Check SSR: theme/twentytwentythree"
                }
            ],
            "endTime": 1698224374269,
            "message": "\\u001b[1m\\u001b[31m  \\u001b[1m● \\u001b[22m\\u001b[1mServer-side Rendering › Check SSR endoint: theme/twentytwentythree › Check SSR: theme/twentytwentythree\\u001b[39m\\u001b[22m\n\n\n\\u001b[2m\\u001b[22m\n\\u001b[2m    \\u001b[0m \\u001b[90m 22 |\\u001b[39m     \\u001b[36mawait\\u001b[39m page\\u001b[33m.\\u001b[39mlocator( \\u001b[32m'#wpcom[data-calypso-ssr=\"true\"]'\\u001b[39m )\\u001b[33m.\\u001b[39mwaitFor( { timeout\\u001b[33m:\\u001b[39m \\u001b[35m15\\u001b[39m \\u001b[33m*\\u001b[39m \\u001b[35m1000\\u001b[39m } )\\u001b[33m;\\u001b[39m\\u001b[0m\\u001b[22m\n\\u001b[2m    \\u001b[0m \\u001b[90m 23 |\\u001b[39m     \\u001b[36mif\\u001b[39m ( endpoint \\u001b[33m===\\u001b[39m \\u001b[32m'theme/twentytwentythree'\\u001b[39m ) {\\u001b[0m\\u001b[22m\n\\u001b[2m    \\u001b[0m\\u001b[31m\\u001b[1m>\\u001b[22m\\u001b[2m\\u001b[39m\\u001b[90m 24 |\\u001b[39m      \\u001b[36mthrow\\u001b[39m \\u001b[36mnew\\u001b[39m \\u001b[33mError\\u001b[39m()\\u001b[33m;\\u001b[39m\\u001b[0m\\u001b[22m\n\\u001b[2m    \\u001b[0m \\u001b[90m    |\\u001b[39m            \\u001b[31m\\u001b[1m^\\u001b[22m\\u001b[2m\\u001b[39m\\u001b[0m\\u001b[22m\n\\u001b[2m    \\u001b[0m \\u001b[90m 25 |\\u001b[39m     }\\u001b[0m\\u001b[22m\n\\u001b[2m    \\u001b[0m \\u001b[90m 26 |\\u001b[39m    } )\\u001b[33m;\\u001b[39m\\u001b[0m\\u001b[22m\n\\u001b[2m    \\u001b[0m \\u001b[90m 27 |\\u001b[39m   }\\u001b[33m,\\u001b[39m\\u001b[0m\\u001b[22m\n\\u001b[2m\\u001b[22m\n\\u001b[2m      \\u001b[2mat Object.<anonymous> (\\u001b[22m\\u001b[2m\\u001b[0m\\u001b[36mspecs/infrastructure/infrastructure__ssr-enabled.ts\\u001b[39m\\u001b[0m\\u001b[2m:24:12)\\u001b[22m\\u001b[2m\\u001b[22m\n",
            "name": "/Users/edwintakahashi/workspace/wp-calypso/test/e2e/specs/infrastructure/infrastructure__ssr-enabled.ts",
            "startTime": 1698224361875,
            "status": "failed",
            "summary": ""
        }
    ],
    "wasInterrupted": false
}

```
</details>